pypreprocess/nipype_preproc_fsl_utils.py

```diff
@@ -180,3 +181,174 @@ def do_subject_preproc(subject_id,
+                           in_index=[2], in_topup_fieldcoef=fieldcoef,
+                           in_topup_movpar=movpar, output_type='NIFTI',
+                           method='jac') for img in func_imgs]
+
+    else:
+        # Merge AP and PA images
+        appa_merged = fsl.Merge(in_files=[ap_realig_img, pa_realig_img],
+                                dimension='t').run()
+
+        # Compute topup transformation
+        acq_param_file = os.path.join('/tmp', 'pypreprocess_topup',
+                                      'acq_param.txt')
+        if not os.path.isfile(acq_param_file):
+            with open(acq_param_file, 'w') as acq:
+                content = '0 -1 0 {0:6.5f} \n0 1 0 {0:6.5f}'
+                acq.write(content % total_readout_times)
+
+        correction = fsl.TOPUP(in_file=appa_merged.outputs.merged_file,
+                               encoding_file=acq_param_file,
+                               output_type='NIFTI',
+                               out_base='APPA_DefMap',
+                               out_field='sanitycheck_DefMap',
+                               out_corrected='sanitycheck_unwarped_B0').run()
+
+        # Apply topup correction to images
+        fieldcoef = correction.outputs.out_fieldcoef
+        movpar = correction.outputs.out_movpar
+        return [fsl.ApplyTOPUP(in_files=img, encoding_file=acq_param_file,
+                               in_index=[2], in_topup_fieldcoef=fieldcoef,
+                               in_topup_movpar=movpar, output_type='NIFTI',
+                               method='jac').run() for img in func_imgs]
+
+
+def _do_subject_topup_correction(subject_data, caching=True,
+                                 hardlink_output=True):
+    """
+    Apply topup correction to subject functional images as implemented in FSL.
+
+    subject_data.topup is expected for any correction to take place. It must
+    contain a dictionary with imgs in subject_data.func as key and a
+    tuple of the form (string, string, (float, float)) corresponding to
+    (ap_img, pa_img, total_readout_times) as value. Its crucial that the ap
+    and pa imgs have already been realigned.
+
+    Parameters
+    ----------
+    subject_data: `SubjectData` object
+        subject data whose functional images are to be corrected with topup
+        (subject_data.func and subject_data.topup)
+
+    caching: bool, optional (default True)
+        if true, then caching will be enabled
+
+    **kwargs:
+       additional parameters to the back-end (SPM, FSL, python)
+
+    Returns
+    -------
+    subject_data: `SubjectData` object
+        preprocessed subject_data. The func and anatomical fields
+        (subject_data.func and subject_data.anat) now contain the
+        oregistered and anatomical images functional images of the subject
+    """
+    corrected_func = []
+    subject_data.nipype_results['topup_correction'] = []
+
+    imgs_to_topup = [img for img in subject_data.func if img in
```

Comment:
This line looks weird to me. What's the difference / relation btw subject_data.func and subject_data.topup ?